Nicer player metadata display

jazzpi opened the pull request:
In the menu view for the players widget:

- show title, artists and album on separate lines
- show album cover (`mpris:artUrl`)

<img width="433" height="441" alt="image" src="https://github.com/user-attachments/assets/b9c4242a-88aa-4654-a67d-c4e666edcb88" />


I added `reqwest` for downloading the album art. This seems to pull in a lot of extra dependencies, maybe we should use something else instead? It can also only handle HTTP(S) URLs. That's all Spotify uses (which I tested with), but I assume local players might use `file://` URLs instead.

I'm also not too sure about the title/artists/album display, it's a bit hard to tell what's what.

For some reason, `nix fmt` also keeps reformatting the `flake.nix`. @kylesferrazza do you maybe understand why?

MalpenZibo commented:
About request I had the same "problem" here #279 .
I didn't found a solution without less dependencies. For now we could leave request, maybe in the future I will try to find something with less dependencies.

jazzpi commented:
Done. I also messed with the widget layout a bit (see updated screenshot).

jazzpi commented on src/modules/media_player.rs:
```diff
@@ -156,6 +228,59 @@ impl MediaPlayer {
         }
     }
 
+    fn check_cover_update(&mut self, data: &[MprisPlayerData]) -> Action {
```
I agree, but I'm struggling a bit with how to integrate it into the architecture. I thought about adding it to the `State::Active` arm of `MprisPlayerService::start_listening()`. But I need to somehow get the `Task` back to `App::update()` so it is executed, right? I can't really figure out how to do that. I'd be grateful for some suggestions :)

romanstingler commented:
@jazzpi will you add the code of the patch into your PR? 
Otherwise, I can also create a PR on top

jazzpi commented:
Okay, I've moved the cover fetching to the `mpris` service. It now spawns futures for cover art fetching and `select!`s from either DBus events or these futures in the `Active` state.

There's still a couple of things to clean up, but besides that I'm running into one major problem at the moment. It looks to me like the `media_player` module does not receive the last event sent by the `mpris` service. When I pause/unpause Spotify to trigger a DBus  metadata event, `media_player` then receives the previous event (e.g. the `CoverFetched` for the song, or even the `MetadataChanged` if the song changed). Maybe this is also related to the other missing-update issues?

I've added some `debug!`s to show this:

```
...
[2026-01-21 15:02:19.398992 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:391: tx: MetadataUpdate: Ok(())
...
[2026-01-21 15:02:19.399493 +01:00] DEBUG [ashell::modules::media_player] src/modules/media_player.rs:257: sub: MetadataChanged
...
[2026-01-21 15:02:19.399661 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:412: tx: CoverFetched(file:///home/jasper/.mozilla/firefox/firefox-mpris/1755_2.png): Ok(())
[2026-01-21 15:02:19.399692 +01:00] DEBUG [ashell::modules::media_player] src/modules/media_player.rs:260: sub: CoverFetched(file:///home/jasper/.mozilla/firefox/firefox-mpris/1755_2.png, _)
...
[2026-01-21 15:02:19.486172 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:412: tx: CoverFetched(https://i.scdn.co/image/ab67616d0000b2739ae41e1296c0349c2a900c4f): Ok(())
[2026-01-21 15:02:19.507073 +01:00] DEBUG [ashell::modules::media_player] src/modules/media_player.rs:65: rx: MetadataChanged
[2026-01-21 15:02:19.507152 +01:00] DEBUG [ashell::modules::media_player] src/modules/media_player.rs:68: rx: CoverFetched(file:///home/jasper/.mozilla/firefox/firefox-mpris/1755_2.png, _)
...
[2026-01-21 15:02:39.739351 +01:00] DEBUG [ashell::modules::media_player] src/modules/media_player.rs:260: sub: CoverFetched(https://i.scdn.co/image/ab67616d0000b2739ae41e1296c0349c2a900c4f, _)
...
[2026-01-21 15:02:39.742526 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:391: tx: MetadataUpdate: Ok(())
[2026-01-21 15:02:39.753005 +01:00] DEBUG [ashell::modules::media_player] src/modules/media_player.rs:68: rx: CoverFetched(https://i.scdn.co/image/ab67616d0000b2739ae41e1296c0349c2a900c4f, _)
...
```

The `tx:` logs are printed after the `mpris` service sends an event to the channel. The `rx:` logs are printed after the `media_player` module receives the event in the `update()` function. The `sub:`  logs are printed from `MediaPlayer::subscription()`, in a `.map()` call on the subscription.

As you can see, on startup the covers for both media players are fetched, and for both the `CoverFetched` event is sent just fine. But the second `CoverFetched` is only received 20 seconds later (after I paused/unpaused Spotify).

I'm at a complete loss why this is happening. Any ideas?

<details>

<summary>Full log</summary>

```
[2026-01-21 15:02:19.262962 +01:00] INFO [ashell::config] src/config.rs:747: Decoding config file "/home/jasper/.config/ashell/config.toml"
[2026-01-21 15:02:19.263742 +01:00] INFO [ashell::config] src/config.rs:753: Config file loaded successfully
[2026-01-21 15:02:19.368961 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:184: Found MPRIS player services: ["org.mpris.MediaPlayer2.firefox.instance_1_27", "org.mpris.MediaPlayer2.spotify"]
[2026-01-21 15:02:19.376150 +01:00] INFO [ashell::services::mpris] src/services/mpris/mod.rs:341: MPRIS player service initialized
[2026-01-21 15:02:19.376870 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:184: Found MPRIS player services: ["org.mpris.MediaPlayer2.firefox.instance_1_27", "org.mpris.MediaPlayer2.spotify"]
[2026-01-21 15:02:19.395817 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:290: Volume changed: Some(0.7341573205157549)
[2026-01-21 15:02:19.396003 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:383: MPRIS player service receive events: [()]
[2026-01-21 15:02:19.396657 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:184: Found MPRIS player services: ["org.mpris.MediaPlayer2.firefox.instance_1_27", "org.mpris.MediaPlayer2.spotify"]
[2026-01-21 15:02:19.398883 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:386: Refreshing MPRIS player data
[2026-01-21 15:02:19.398992 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:391: tx: MetadataUpdate: Ok(())
libEGL warning: pci id for fd 57: 10de:2486, driver (null)

[2026-01-21 15:02:19.399493 +01:00] DEBUG [ashell::modules::media_player] src/modules/media_player.rs:257: sub: MetadataChanged
[2026-01-21 15:02:19.399607 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:409: Fetched cover art from file:///home/jasper/.mozilla/firefox/firefox-mpris/1755_2.png
[2026-01-21 15:02:19.399661 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:412: tx: CoverFetched(file:///home/jasper/.mozilla/firefox/firefox-mpris/1755_2.png): Ok(())
[2026-01-21 15:02:19.399692 +01:00] DEBUG [ashell::modules::media_player] src/modules/media_player.rs:260: sub: CoverFetched(file:///home/jasper/.mozilla/firefox/firefox-mpris/1755_2.png, _)
pci id for fd 56: 10de:2486, driver (null)
pci id for fd 59: 10de:2486, driver (null)
libEGL warning: egl: failed to create dri2 screen
pci id for fd 56: 10de:2486, driver (null)
pci id for fd 59: 10de:2486, driver (null)
libEGL warning: egl: failed to create dri2 screen
libEGL warning: pci id for fd 62: 10de:2486, driver (null)

pci id for fd 44: 10de:2486, driver (null)
pci id for fd 63: 10de:2486, driver (null)
libEGL warning: egl: failed to create dri2 screen
pci id for fd 44: 10de:2486, driver (null)
pci id for fd 63: 10de:2486, driver (null)
libEGL warning: egl: failed to create dri2 screen
[2026-01-21 15:02:19.485974 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:409: Fetched cover art from https://i.scdn.co/image/ab67616d0000b2739ae41e1296c0349c2a900c4f
[2026-01-21 15:02:19.486172 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:412: tx: CoverFetched(https://i.scdn.co/image/ab67616d0000b2739ae41e1296c0349c2a900c4f): Ok(())
[2026-01-21 15:02:19.507073 +01:00] DEBUG [ashell::modules::media_player] src/modules/media_player.rs:65: rx: MetadataChanged
[2026-01-21 15:02:19.507152 +01:00] DEBUG [ashell::modules::media_player] src/modules/media_player.rs:68: rx: CoverFetched(file:///home/jasper/.mozilla/firefox/firefox-mpris/1755_2.png, _)
[2026-01-21 15:02:19.507186 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:141: Updating MPRIS player cover for file:///home/jasper/.mozilla/firefox/firefox-mpris/1755_2.png, 71164 bytes
[2026-01-21 15:02:39.739066 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:311: PlaybackStatus changed: Playing
[2026-01-21 15:02:39.739172 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:383: MPRIS player service receive events: [()]
[2026-01-21 15:02:39.739351 +01:00] DEBUG [ashell::modules::media_player] src/modules/media_player.rs:260: sub: CoverFetched(https://i.scdn.co/image/ab67616d0000b2739ae41e1296c0349c2a900c4f, _)
[2026-01-21 15:02:39.740012 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:184: Found MPRIS player services: ["org.mpris.MediaPlayer2.firefox.instance_1_27", "org.mpris.MediaPlayer2.spotify"]
[2026-01-21 15:02:39.742438 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:386: Refreshing MPRIS player data
[2026-01-21 15:02:39.742526 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:391: tx: MetadataUpdate: Ok(())
[2026-01-21 15:02:39.753005 +01:00] DEBUG [ashell::modules::media_player] src/modules/media_player.rs:68: rx: CoverFetched(https://i.scdn.co/image/ab67616d0000b2739ae41e1296c0349c2a900c4f, _)
[2026-01-21 15:02:39.753114 +01:00] DEBUG [ashell::services::mpris] src/services/mpris/mod.rs:141: Updating MPRIS player cover for https://i.scdn.co/image/ab67616d0000b2739ae41e1296c0349c2a900c4f, 86294 bytes
```

</details>

romanstingler commented on src/services/mpris/mod.rs:
```diff
@@ -277,68 +322,169 @@ impl MprisPlayerService {
-                            Ok(data) => {
-                                debug!("Refreshing MPRIS player data");
+                let mut pending_downloads = FuturesUnordered::new();
+                let mut in_flight: HashMap<String, AbortHandle> = HashMap::new();
```
What about keeping both structures inside the `State::Active` variant, so when the loop exits we hand the existing futures back into the next `active()` run and nothing is lost.
We can also treat a `None` chunk as a signal to transition to `Init` or `Error` and rebuild the connection if needed, but at minimum persisting the state prevents dropped downloads.
maybe something like 
[mpris-state-persistence.patch](https://github.com/user-attachments/files/24779395/mpris-state-persistence.patch)

romanstingler commented on src/modules/media_player.rs:
```diff
@@ -103,22 +133,52 @@ impl MediaPlayer {
+                                    image(image::Handle::from_bytes(img.clone()))
+                                        .filter_method(image::FilterMethod::Linear)
+                                        .width(buttons_width)
+                                        .into()
```
I agree that cloning on every menu_view render is wasteful, but I’d rather keep the MPRIS service free of iced types. 
I think we can add a small cache inside the media module (`cover_handles: HashMap<String, image::Handle>`) that syncs with `MetadataChanged/CoverFetched `events. 
The service still owns plain Bytes, and the module lazily wraps them into handles once per URL, pruning any that fall out of the metadata. 
This gives us the reuse we wanted without pushing UI concerns into the service. 
Maybe something like 
[cover-handle-cache.patch](https://github.com/user-attachments/files/24778648/cover-handle-cache.patch)

romanstingler commented:
> I replaced `pending_downloads.next().fuse()` with `pending_downloads.select_next_some()` and that fixed it. Now ashell also doesn't use a full CPU core anymore 🙈

probably you built-in a crypto miner :P

jazzpi commented:
@romanstingler that's a different issue from @MalpenZibo's (since there, there is metadata, just no cover). I just tested with the current nixpkgs version (i.e. 0.7.0) and the issue also happens there. So I don't think it's related to the changes in this PR, it's part of #447.

MalpenZibo commented:
<img width="872" height="546" alt="image" src="https://github.com/user-attachments/assets/7a99ea63-65f7-4a03-9887-8d1783cd5090" />
Nice, it seems to be working fine. I'll do a final review, and then we're ready to merge it

clotodex commented:
Tag me when it is ready for testing then i will try it out a bit
